File: src/me/memorial/module/modules/client/Notifications.java
package me.memorial.module.modules.client;

import lombok.Getter;
import me.memorial.events.EventTarget;
import me.memorial.events.impl.render.Render2DEvent;
import me.memorial.module.Module;
import me.memorial.module.ModuleCategory;
import me.memorial.module.ModuleInfo;
import me.memorial.ui.font.Fonts;
import me.memorial.utils.render.RenderUtils;
import me.memorial.value.IntegerValue;
import me.memorial.value.ListValue;
import net.minecraft.client.gui.FontRenderer;

import java.awt.*;

@ModuleInfo(name = "Notifications",description = "notif", category = ModuleCategory.CLIENT)
public class Notifications extends Module {
    private final ListValue mode = new ListValue("Mode", new String[]{"test","test2","test3","none"},"test");
    public IntegerValue x = new IntegerValue("X", 110, 1, 120);
    public IntegerValue y = new IntegerValue("Y", 20, 1,120);

    @Getter
    public static Notifications instance;

    @Getter
    private float width = 0.0f;
    @Getter
    private float height = 0.0f;
    @Getter
    private float x1 = x.get();


    public Notifications() {
        instance = this;
    }

    public void add(String str, FontRenderer font, Color rectcolor,Color fontcolor, boolean shadowtext) {

        if(mode.get().equals("none")) {
            return;
        }

        width = font.getStringWidth(str);
        height = font.FONT_HEIGHT;

        drawnotif(rectcolor);

        /*
        if (shadowtext) {
            font.drawStringWithShadow(str, x1 , y.get() , fontcolor.getRGB());
        } else {
            font.drawString(str, (int) x1, y.get() , fontcolor.getRGB());
        }

         */

        Fonts.font30.drawString(str, 7F, 17F, -1);

    }

    @EventTarget
    void onRender2D(Render2DEvent event) {
        add("test",Fonts.font35,Color.BLACK,Color.WHITE,true);
    }


    private void drawnotif(Color color) {

        if (mode.get().equals("test")) {
            RenderUtils.drawRoundedRect(x1, y.get(), width, height, 0f, color);
        }

        if (mode.get().equals("test2")) {
            RenderUtils.drawRect(x1, y.get(), width, height, color);
        }

        if(mode.get().equals("test3")) {
            RenderUtils.drawRect(22F, 0F, width, height, color);
            Fonts.font35.drawString("Module", 7F, 4F, -1);
        }

        RenderUtils.resetColor();

    }


}

File: src/me/memorial/module/modules/client/Notifications.kt
package me.memorial.module.modules.client

import me.memorial.module.Module
import me.memorial.module.ModuleCategory
import me.memorial.module.ModuleInfo
import me.memorial.module.modules.client.FadeState.*
import me.memorial.ui.font.Fonts
import me.memorial.utils.ClientUtils
import me.memorial.utils.EaseUtils
import me.memorial.utils.render.RenderUtils
import me.memorial.value.BoolValue
import org.lwjgl.opengl.GL11
import java.awt.Color
import kotlin.math.max

@ModuleInfo(name = "Notifications", description = "notif", category = ModuleCategory.CLIENT)
class Notifications : Module() {

    val winnotif = BoolValue("winnotif",false)
    val lognotif = BoolValue("lognotif",true)

     fun add(text: String,color: Color) {
        val width =  Fonts.font30.getStringWidth(text).toFloat()
        val height = Fonts.font30.height
        val animeTime = 500
        val nowTime = System.currentTimeMillis()
        val displayTime = System.currentTimeMillis()
        var nowY = -height
        val time = 1500
        val realY = -(0 + 1) * height
        var transY = nowY.toDouble()
        var animeXTime = System.currentTimeMillis()
        var animeYTime = System.currentTimeMillis()
        var fadeState = IN

        me.memorial.utils.ClientUtils.loginfo("有操作好像不显示")
         me.memorial.utils.ClientUtils.loginfo("有操作好像不显示")
         me.memorial.utils.ClientUtils.loginfo("有操作好像不显示")
         me.memorial.utils.ClientUtils.loginfo("有操作好像不显示")
         me.memorial.utils.ClientUtils.loginfo("有操作好像不显示")
         me.memorial.utils.ClientUtils.loginfo("有操作好像不显示")


         // Y-Axis Animation
        if (nowY != realY) {
            var pct = (nowTime - animeYTime) / animeTime.toDouble()
            if (pct > 1) {
                nowY = realY
                pct = 1.0
            } else {
                pct = EaseUtils.easeOutExpo(pct)
            }
            transY += (realY - nowY) * pct
        } else {
            animeYTime = nowTime
        }

        // X-Axis Animation
        var pct = (nowTime - animeXTime) / animeTime.toDouble()
        when (fadeState) {
            IN -> {
                if (pct > 1) {
                    fadeState = STAY
                    animeXTime = nowTime
                    pct = 1.0
                }
                pct = EaseUtils.easeOutExpo(pct)
            }

            STAY -> {
                pct = 1.0
                if ((nowTime - animeXTime) > time) {
                    fadeState = OUT
                    animeXTime = nowTime
                }
            }

            OUT -> {
                if (pct > 1) {
                    fadeState = END
                    animeXTime = nowTime
                    pct = 1.0
                }
                pct = 1 - EaseUtils.easeInExpo(pct)
            }

            END -> TODO()
        }

         me.memorial.utils.ClientUtils.loginfo("有操作好像不显示3")


         val transX = width - (width * pct) - width
        GL11.glTranslated(transX, transY, 0.0)

         me.memorial.utils.ClientUtils.loginfo("有操作好像不显示4")

         RenderUtils.drawRoundedCornerRect(
            0F + 3f,
            0F,
            max(width - width * ((nowTime - displayTime) / (animeTime * 2F + time)) + 5f, 0F),
            27f - 5f,
            2f,
            Color(0, 0, 0, 150).rgb
        )
         me.memorial.utils.ClientUtils.loginfo("有操作好像不显示5")

         // if (state) Color(0x60E092).rgb else Color(0xFF2F2F).rgb
        RenderUtils.drawRoundedCornerRect(3F, 0F, width + 5f, 27f - 5f, 2f, color.rgb)
         me.memorial.utils.ClientUtils.loginfo("有操作好像不显示6")

         me.memorial.utils.ClientUtils.loginfo("有操作好像不显示6 获取文本$text")

         Fonts.font35.drawCenteredString("Module", 25F, 3F, -1)
         me.memorial.utils.ClientUtils.loginfo("有操作好像不显示7")

         Fonts.font30.drawCenteredString(text,30F, 12F, -1)
         me.memorial.utils.ClientUtils.loginfo("有操作好像不显示8")

     }


}

